authentication/php/phpleague/src/PhpLeagueAccessTokenProvider.php

```diff
@@ -0,0 +1,115 @@
+<?php
+/**
+ * Copyright (c) Microsoft Corporation.  All Rights Reserved.
+ * Licensed under the MIT License.  See License in the project root
+ * for license information.
+ */
+
+
+namespace Microsoft\Kiota\Authentication;
+
+
+use Http\Promise\FulfilledPromise;
+use Http\Promise\Promise;
+use Http\Promise\RejectedPromise;
+use League\OAuth2\Client\Provider\GenericProvider;
+use League\OAuth2\Client\Token\AccessToken;
+use Microsoft\Kiota\Abstractions\Authentication\AccessTokenProvider;
+use Microsoft\Kiota\Abstractions\Authentication\AllowedHostsValidator;
+
+/**
+ * Class PhpLeagueAccessTokenProvider
+ * @package Microsoft\Kiota\Authentication
+ * @copyright 2022 Microsoft Corporation
+ * @license https://opensource.org/licenses/MIT MIT License
+ * @link https://developer.microsoft.com/graph
+ */
+class PhpLeagueAccessTokenProvider implements AccessTokenProvider
+{
+    /**
+     * @var TokenRequestContext {@link TokenRequestContext}
+     */
+    private TokenRequestContext $tokenRequestContext;
+    /**
+     * @var AllowedHostsValidator Validates whether a token should be fetched for a request url
+     */
+    private AllowedHostsValidator $allowedHostsValidator;
+    /**
+     * @var array<string, string>
+     */
+    private array $scopes;
+    /**
+     * @var AccessToken|null Token object to re-use before expiry
+     */
+    private ?AccessToken $cachedToken = null;
+    /**
+     * @var GenericProvider OAuth 2.0 provider from PHP League library
+     */
+    private GenericProvider $oauthProvider;
+
+    /**
+     * Creates a new instance
+     * @param TokenRequestContext $tokenRequestContext
+     * @param array $scopes
+     * @param array $allowedHosts
+     */
+    public function __construct(TokenRequestContext $tokenRequestContext, array $scopes, array $allowedHosts = [])
+    {
+
+        $this->tokenRequestContext = $tokenRequestContext;
+        if (empty($scopes)) {
+            throw new \InvalidArgumentException("Scopes cannot be empty");
+        }
+        $this->scopes = $scopes;
+        $this->allowedHostsValidator = new AllowedHostsValidator($allowedHosts);
+        $this->oauthProvider = new GenericProvider([
+            'urlAccessToken' => "https://login.microsoftonline.com/{$tokenRequestContext->getTenantId()}/oauth2/v2.0/token",
+            'urlAuthorize' => '', // required to instantiate Generic Provider
+            'urlResourceOwnerDetails' => '' // required to instantiate Generic Provider
```

Comment:
The lib doesn't support sign out...
But I also don't see the need to support sign out.

I thought the authentication providers are only concerned with getting access tokens meaning all prior steps to the access token request e.g. requesting auth codes/id tokens etc will need to be done before initializing this auth provider?

This way the dev can set their preferred redirect uri & once Identity redirects to their application with the auth code, they can init the auth provider & ApiClient etc?

Please correct me if I'm wrong.

Reply:
from a sanity perspective and a security standpoint, if a user clicks "sign out" somewhere in a UI to make sure the app doesn't have further access to their data until they interact with it again.
This means a couple of things:
- cache invalidation, the app "doesn't know" how to access the data anymore because it doesn't have the tokens anymore
- tokens invalidation: should the token have been compromised somehow, or still be in cache, they have been invalidated (this goes hand in hand with CAE and other things)
- session invalidation: the identity provider will request a sign-in flow again should it get another token request from the same "context".

signing out through OIDC supports the later, and maybe even the second point through inner mechanics of the identity platform.

Reply:
There's only one OIDC lib that supports most of the features like sign-out, token validation, introspection etc. Unfortunately it's not yet stable (v0.x). For now I'll stick to PHP League's lib.
There's more time allocated next quarter for PHP Auth. We could consider implementing OIDC capabilities. Created an issue to track this (https://github.com/microsoft/kiota/issues/1238)